samza-core/src/main/scala/org/apache/samza/container/TaskInstance.scala

```diff
@@ -150,20 +162,28 @@ class TaskInstance(
     })
   }
 
-  def process(envelope: IncomingMessageEnvelope, coordinator: ReadableCoordinator,
+  def process(msgEnvelope: IncomingMessageEnvelope, coordinator: ReadableCoordinator,
     callbackFactory: TaskCallbackFactory = null) {
     metrics.processes.inc
 
-    if (!ssp2catchedupMapping.getOrElse(envelope.getSystemStreamPartition,
-      throw new SamzaException(envelope.getSystemStreamPartition + " is not registered!"))) {
-      checkCaughtUp(envelope)
+    if (!ssp2catchedupMapping.getOrElse(msgEnvelope.getSystemStreamPartition,
+      throw new SamzaException(msgEnvelope.getSystemStreamPartition + " is not registered!"))) {
+      checkCaughtUp(msgEnvelope)
     }
 
-    if (ssp2catchedupMapping(envelope.getSystemStreamPartition)) {
+    if (ssp2catchedupMapping(msgEnvelope.getSystemStreamPartition)) {
       metrics.messagesActuallyProcessed.inc
 
       trace("Processing incoming message envelope for taskName and SSP: %s, %s"
-        format (taskName, envelope.getSystemStreamPartition))
+        format (taskName, msgEnvelope.getSystemStreamPartition))
+
+      val envelope: IncomingMessageEnvelope = if (msgEnvelope.getMessage.isInstanceOf[ControlMessage]) {
+        controlManager.update(msgEnvelope)
```

Comment:
This is not conforming to what we discussed earlier in the SEP. I thought that the interception of control messages and the reconciliation / propagation of control message would be done in TaskInstance. It seems that it is now separated in two pieces: update the EoS status of each stream in TaskInstance, and propagate EoS in StreamOperatorTask. Is that the current implementation? Can we elaborate why not do everything in TaskInstance? And even we decided that propagate EoS is not a general function for all task classes in TaskInstance, I would prefer to add a new interface ControlMessageListnerTask to handle onControlMessage(ControlMessage msg), instead of borrowing the IncomingMessageEnvelope which is associated to a single incoming SSP, which the consolidated control message apparently is meant for *all partitions* for the stream to a single task.